product_supplier_sale_price/models/product.py

```diff
@@ -0,0 +1,148 @@
+            product_vals = product._prepare_supplier_sale_price()
+            product.write(product_vals)
+
+    @api.model
+    def create(self, vals):
+        new_supplierinfo = super(ProductSupplierinfo, self).create(vals)
+        if vals.get('sale_price'):
+            new_supplierinfo._update_supplier_sale_price()
+        return new_supplierinfo
+
+    @api.multi
+    def write(self, vals):
+        res = super(ProductSupplierinfo, self).write(vals)
+
+        if vals.get('sale_price'):
+            for supplierinfo in self:
+                supplierinfo._update_supplier_sale_price()
+
+        return res
+
+
+class ProductProduct(models.Model):
+    _inherit = 'product.product'
+
+    @api.multi
+    @api.depends('seller_ids.sale_price', 'use_supplier_sale_price')
+    def _compute_supplier_sale_price(self):
+        param_obj = self.env['ir.config_parameter']
+
+        sale_price_computation = param_obj.sudo().get_param(
+            'product_supplier_sale_price'
+            '.field_res_config_settings_'
+            '_product_supplier_sale_price_computation')
+
+        for product in self:
+            supplier_prices = []
+            for seller in product.variant_seller_ids:
+                if seller.sale_price\
+                        and (seller.product_id.default_code ==
+                             product.default_code
+                             or (not seller.product_id and
+                                 seller.product_tmpl_id ==
+                                 product.product_tmpl_id)):
+                    supplier_prices.append(seller.sale_price)
+            if not supplier_prices:
+                # this product has no supplier sale prices
+                product.supplier_sale_price = 0.0
+                continue
+            if sale_price_computation == 'more_expensive_sale_price':
+                sale_price = max(supplier_prices)
+            elif sale_price_computation == 'cheapest_sale_price':
+                sale_price = min(supplier_prices)
+            elif sale_price_computation == 'mean_sale_price':
+                sale_price = statistics.mean(supplier_prices)
+            else:  # 'not_supplier_sale_price'
+                sale_price = False
+            product.supplier_sale_price = sale_price
+
+    use_supplier_sale_price = fields.Boolean('Use Supplier Sale Price')
+
+    supplier_sale_price = fields.Float(
+        'Supplier_sale_price', compute='_compute_supplier_sale_price',
+        readonly=True)
+
+    @api.multi
+    @api.depends('fix_price', 'use_supplier_sale_price',
+                 'variant_seller_ids.sale_price')
+    def _compute_lst_price(self):
+        for product in self:
+            if product.use_supplier_sale_price:
+                product.lst_price = product.supplier_sale_price
+            else:
+                super(ProductProduct, product)._compute_lst_price()
+
+    @api.multi
+    def _prepare_supplier_sale_price(self, force=False):
```

Comment:
I would have done this implementing multi recordset in order to improve performances.